parquet-column/src/main/java/org/apache/parquet/column/impl/ColumnWriterBase.java

```diff
@@ -97,7 +97,7 @@ abstract class ColumnWriterBase implements ColumnWriter {
       int optimalNumOfBits = BlockSplitBloomFilter.optimalNumOfBits(ndv.getAsLong(), fpp.getAsDouble());
       this.bloomFilter = new BlockSplitBloomFilter(optimalNumOfBits / 8, maxBloomFilterSize);
     } else {
-      this.bloomFilter = new BlockSplitBloomFilter(maxBloomFilterSize);
+      this.bloomFilter = BlockSplitBloomFilter.of(maxBloomFilterSize);
```

Comment:
```suggestion
      this.bloomFilter = new BlockSplitBloomFilter(maxBloomFilterSize, maxBloomFilterSize);
```

Reply:
Does this change solve the problem?

Reply:
Yes, but it guarantees the maximum bytes size, but it cannot be guaranteed to be a power of 2. I am afraid that there will be a problem here.

Reply:
https://github.com/apache/parquet-mr/blob/master/parquet-column/src/main/java/org/apache/parquet/column/values/bloomfilter/BlockSplitBloomFilter.java#L146

It tries to find the next power of two and also guarantees not to exceed the max size.

BTW, the specs does not require the size to be power of two. Is there any issue you have seen if the size is not power of two?

Reply:
If I understand something wrong, please correct me. I found this part of code, need `bitset.length / BYTES_PER_BLOCK` result to be integer, need power of 2, extra bit is useless, but also no error.
https://github.com/apache/parquet-mr/blob/1235003e742e6a76bf6cb8f7ed33e942fa12d0d5/parquet-column/src/main/java/org/apache/parquet/column/values/bloomfilter/BlockSplitBloomFilter.java#L245-L258

Reply:
OK, I got your point. According to the delegate constructor below, `LOWER_BOUND_BYTES` and `UPPER_BOUND_BYTES` are hard limits. What about checking if `minimumBytes` and `maximumBytes` are power of two and throw if not?

```java
  public BlockSplitBloomFilter(int numBytes, int minimumBytes, int maximumBytes, HashStrategy hashStrategy) {
    if (minimumBytes > maximumBytes) {
      throw new IllegalArgumentException("the minimum bytes should be less or equal than maximum bytes");
    }

    if (minimumBytes > LOWER_BOUND_BYTES && minimumBytes < UPPER_BOUND_BYTES) {
      this.minimumBytes = minimumBytes;
    }

    if (maximumBytes > LOWER_BOUND_BYTES && maximumBytes < UPPER_BOUND_BYTES) {
      this.maximumBytes = maximumBytes;
    }

    initBitset(numBytes);

    cacheBuffer.order(ByteOrder.LITTLE_ENDIAN);

    switch (hashStrategy) {
      case XXH64:
        this.hashStrategy = hashStrategy;
        hashFunction = new XxHash();
        break;
      default:
        throw new RuntimeException("Unsupported hash strategy");
    }
  }
```

Reply:
I fear there will be forward compatibility issues if throw exception when they are not power of two.
Actually here if the bitSet size is not a power of 2, there will be no error, but some storage is wasted.

Reply:
I think that is fine. It is well explained in the comment of the constructor. WDYT @gszadovszky

Reply:
> If I understand something wrong, please correct me. I found this part of code, need `bitset.length / BYTES_PER_BLOCK` result to be integer, need power of 2, extra bit is useless, but also no error.

@wgtmac @ggershinsky Sorry, I may have misunderstood. `bitset.length` is valid as long as it is a multiple of `BYTES_PER_BLOCK`(32) (not necessarily a power of 2) . 
The reason for the power of 2 here is more complicated, I think there are performance reasons. 
Most of the existing parquet uses need to specify the NDV value or default maxBytes 1MB. In these cases, `numbytes` will be always initialized to a power of 2.
@chenjunjiedada could you take a look at this? thank you.

Reply:
> If I understand something wrong, please correct me. I found this part of code, need `bitset.length / BYTES_PER_BLOCK` result to be integer, need power of 2, extra bit is useless, but also no error.

@wgtmac @gszadovszky  Sorry, I may have misunderstood. `bitset.length` is valid as long as it is a multiple of `BYTES_PER_BLOCK`(32) (not necessarily a power of 2) . 
The reason for the power of 2 here is more complicated, I think there are performance reasons. 
Most of the existing parquet uses need to specify the NDV value or default maxBytes 1MB. In these cases, `numbytes` will be always initialized to a power of 2.
@chenjunjiedada could you take a look at this? thank you.

Reply:
I refer to the documentation here https://github.com/apache/parquet-format/blob/master/BloomFilter.md#technical-approach and learned that the implementation in parquet is based on the paper http://algo2.iti.kit.edu/documents/cacheefficientbloomfilters-jea.pdf 
The power of 2 is probably for performance requirements 

Impala has the same implementation but i'm not familiar with it... I saw similar implementation logic here.
https://github.com/apache/impala/blob/2c779939dc302be9ee5dd97ddf374bb043040891/be/src/kudu/util/block_bloom_filter.h#L88-L92

Reply:
Yes, I agree that power of two is solely for best performance.

IMHO, fixing the delegate constructor is the cleanest way to address this. I think it is OK to break users if the actual size exceeds `parquet.bloom.filter.max.bytes` as this is a bug.

Reply:
I agree with @wgtmac. Fixing a defect might lead to breaking compatibility.

Reply:
@wgtmac @gszadovszky sorry, I didn't understand the meaning, do you mean to use `BlockSplitBloomFilter(int numBytes, int maximumBytes)` API (even if `numBytes` is not power of 2)?

Reply:
The goal is to fix the issue that bloom filter size can be larger than `parquet.bloom.filter.max.bytes`. The problematic constructors used are as below:

```java
    int maxBloomFilterSize = props.getMaxBloomFilterBytes();

    OptionalLong ndv = props.getBloomFilterNDV(path);
    OptionalDouble fpp = props.getBloomFilterFPP(path);
    // If user specify the column NDV, we construct Bloom filter from it.
    if (ndv.isPresent()) {
      int optimalNumOfBits = BlockSplitBloomFilter.optimalNumOfBits(ndv.getAsLong(), fpp.getAsDouble());
      this.bloomFilter = new BlockSplitBloomFilter(optimalNumOfBits / 8, maxBloomFilterSize);
    } else {
      this.bloomFilter = new BlockSplitBloomFilter(maxBloomFilterSize);
    }
```

They are all delegated by `public BlockSplitBloomFilter(int numBytes, int minimumBytes, int maximumBytes, HashStrategy hashStrategy)`. 
```java
  public BlockSplitBloomFilter(int numBytes) {
    this(numBytes, LOWER_BOUND_BYTES, UPPER_BOUND_BYTES, HashStrategy.XXH64);
  }

  public BlockSplitBloomFilter(int numBytes, int maximumBytes) {
    this(numBytes, LOWER_BOUND_BYTES, maximumBytes, HashStrategy.XXH64);
  }
```

So it seems to me that the issue comes from `public BlockSplitBloomFilter(int numBytes)` where `maxBloomFilterSize` is passed to `numBytes` and the actual size can be adjusted to values as large as `UPPER_BOUND_BYTES`.

Is it possible to fix it like below?
```java
  public BlockSplitBloomFilter(int numBytes) {
    this(numBytes, LOWER_BOUND_BYTES, numBytes, HashStrategy.XXH64);
  }
```

Reply:
@wgtmac Thank you for the clear explanation! 
I worry that if we modify it like this, it will reduces the ability of one constructor, that is, the initial size is `numbytes`, but the maximum does not exceed `UPPER_BOUND_BYTES`.
In fact, I think it might be a problem of calling the constructor incorrectly.  As you said before, we can change to call `BlockSplitBloomFilter(int numBytes, int maximumBytes)` instead of `BlockSplitBloomFilter(int numBytes)`. 
What do you think about it?

Reply:
I agree with you. @yabola